examples/cmd/testing/recordingscript.go

```diff
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"flag"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/src-d/metadata-retrieval/github"
+
+	"github.com/motemen/go-loghttp"
+	"golang.org/x/oauth2"
+	"gopkg.in/src-d/go-log.v1"
+)
+
+// This script will use the downloader to crawl using the graphql API a repository (default: src-d/gitbase)
+// and an organization (default: src-d) and will record requests (graphql queries) and responses (json body data)
+// into a map stored in a gob file.
+// NB: script is assumed to run from the project root (for correct file storage)
+func main() {
+
+	var (
+		org  string
+		repo string
+	)
+
+	flag.StringVar(&org, "org", "src-d", "a GitHub organization")
+	flag.StringVar(&repo, "repo", "gitbase", "a GitHub repository")
+	flag.Parse()
+
+	// Variables to hold graphql queries (keys) and responses (values)
+	var (
+		reqResp map[string]string = make(map[string]string)
+		query   string
+	)
+
+	// Create a with authentication and add logging (recording)
+	client := oauth2.NewClient(
+		context.TODO(),
+		oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: os.Getenv("GITHUB_TOKEN")},
+		))
+
+	client.Transport = &loghttp.Transport{
+		Transport: client.Transport,
+		LogRequest: func(req *http.Request) {
+			// consume request body
+			savecl := req.ContentLength
+			bodyBytes, _ := ioutil.ReadAll(req.Body)
+			defer req.Body.Close()
+			// recreate request body
+			req.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
+			req.ContentLength = savecl
+			query = string(bodyBytes)
+		},
+		LogResponse: func(resp *http.Response) {
+			// TODO(@kyrcha): also record other types of responses
+			if resp.StatusCode == http.StatusOK {
+				// consume response body
+				savecl := resp.ContentLength
+				bodyBytes, _ := ioutil.ReadAll(resp.Body)
+				defer resp.Body.Close()
+				// recreate response body
+				resp.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
+				resp.ContentLength = savecl
+				// save response body
+				reqResp[query] = string(bodyBytes)
+			}
+		},
+	}
+
+	downloader, err := github.NewStdoutDownloader(client)
+	if err != nil {
+		panic(err)
+	}
+
+	// record a repo crawl
+	log.Infof("Start recording a repo")
+	downloader.DownloadRepository(context.TODO(), org, repo, 0)
```

Comment:
here and line 44 `context` is required, maybe it would be more useful to create one at the beginning and pass it to functions?